src/orange/detail/BIHMaker.cc

```diff
@@ -0,0 +1,329 @@
+        auto bbox = bboxes_[indices[i].unchecked_get()];
+
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            center[ax] = (bbox.lower()[ax] + bbox.upper()[ax]) / 2;
+        }
+        centers[i] = center;
+    }
+    return centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create sorted and uniquified X, Y, Z values of bbox centers.
+ */
+BIHMaker::AxesCenters BIHMaker::axes_centers(VecReal3 const& centers)
+{
+    CELER_EXPECT(!centers.empty());
+
+    AxesCenters axes_centers{{}, {}, {}};
+
+    for (auto center : centers)
+    {
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            axes_centers[ax].push_back(center[ax]);
+        }
+    }
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        sort_and_uniquify(axes_centers[ax]);
+    }
+
+    return axes_centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Bounding box of a collection of bounding boxes.
+ */
+BoundingBox BIHMaker::meta_bbox(VecIndices const& indices)
+{
+    CELER_EXPECT(!indices.empty());
+
+    auto inf = std::numeric_limits<double>::infinity();
+
+    Real3 lower = {inf, inf, inf};
+    Real3 upper = {-inf, -inf, -inf};
+
+    for (auto id : indices)
+    {
+        auto bbox = bboxes_[id.unchecked_get()];
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            lower[ax] = std::min(lower[ax], bbox.lower()[ax]);
+            upper[ax] = std::max(upper[ax], bbox.upper()[ax]);
+        }
+    }
+
+    return {lower, upper};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a vector of axes sorted from longest to shortest.
+ */
+BIHMaker::VecAxes BIHMaker::sort_axes(BoundingBox const& bbox)
+{
+    VecAxes axes;
+    std::vector<double> lengths;
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        axes.push_back(axis);
+        lengths.push_back(bbox.upper()[ax] - bbox.lower()[ax]);
+    }
+
+    std::sort(axes.begin(), axes.end(), [&](Axis axis1, Axis axis2) {
+        return lengths[to_int(axis1)] > lengths[to_int(axis2)];
+    });
+    return axes;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Check that only the first bounding box (i.e. exterior volume) is fully inf.
+ */
+bool BIHMaker::check_bbox_extents()
+{
+    return this->fully_inf(bboxes_[0])
+           && std::all_of(
+               bboxes_.begin() + 1, bboxes_.end(), [this](BoundingBox bbox) {
+                   return !this->fully_inf(bbox);
+               });
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Check if a bounding box spans (-inf, inf) in every direction.
+ */
+bool BIHMaker::fully_inf(BoundingBox const& bbox)
```

Comment:
Another bbox utils candidate `is_infinite()` (which is symmetric with `BoundingBox::from_infinite`: if we want semi-infinite detection we could use `is_semi_infinite` or `is_volume_infinite`).